private/pkg/git/git.go

```diff
@@ -133,3 +139,133 @@ type ListFilesAndUnstagedFilesOptions struct {
 	// is being applied to.
 	IgnorePathRegexps []*regexp.Regexp
 }
+
+// BranchIterator ranges over branches for a git repository.
+//
+// Only branches from the remote named `origin` are ranged.
+type BranchIterator interface {
+	// ForEachBranch ranges over branches in the repository in an undefined order.
+	ForEachBranch(func(branch string) error) error
+}
+
+// CommitIterator ranges over commits for a git repository.
+//
+// Only commits from the remote named `origin` are ranged.
+type CommitIterator interface {
+	// BaseBranch is the base branch of the repository. This is either
+	// configured via the `WithBaseBranch` option, or discovered via the
+	// remote named `origin`. Therefore, discovery requires that the repository
+	// is pushed to the remote.
+	BaseBranch() string
+	// ForEachCommit ranges over commits for the target branch in reverse topological order.
+	//
+	// Parents are visited before children, and only left parents are visited (i.e.,
+	// commits from branches merged into the target branch are not visited).
+	ForEachCommit(branch string, f func(commit Commit) error) error
+}
+
+type CommitIteratorOption func(*commitIteratorOpts) error
+
+// CommitIteratorWithBaseBranch configures the base branch for this iterator.
+func CommitIteratorWithBaseBranch(name string) CommitIteratorOption {
+	return func(r *commitIteratorOpts) error {
+		if name == "" {
+			return errors.New("base branch cannot be empty")
+		}
+		r.baseBranch = name
+		return nil
+	}
+}
+
+// NewBranchIterator creates a new BranchIterator that can range over branches.
+func NewBranchIterator(
+	gitDirPath string,
+	objectReader ObjectReader,
+) (BranchIterator, error) {
+	return newBranchIterator(gitDirPath, objectReader)
+}
+
+// NewCommitIterator creates a new CommitIterator that can range over commits.
+//
+// By default, NewCommitIterator will attempt to detect the base branch if the repository
+// has been pushed. This may fail. TODO: we probably want to remove this and
+// force the use of the `WithBaseBranch` option.
+func NewCommitIterator(
+	gitDirPath string,
+	objectReader ObjectReader,
+	options ...CommitIteratorOption,
+) (CommitIterator, error) {
+	var opts commitIteratorOpts
+	for _, option := range options {
+		if err := option(&opts); err != nil {
+			return nil, err
+		}
+	}
+	return newCommitIterator(gitDirPath, objectReader, opts)
+}
+
+// Hash represents the hash of a Git object (tree, blob, or commit).
+type Hash interface {
+	// Hex is the hexadecimal representation of this ID.
+	Hex() string
+	// String returns the hexadecimal representation of this ID.
+	String() string
```

Comment:
Why two funcs for the same information?

Reply:
`String()` is only here for `Stringer` compliance so that it's easier to print.